### ISSUE TITLE
关于数据预处理和finetuning的一些问题

### ISSUE THREAD

**Issue by ChenHu-ML** [935963004/LaBraM#45]
近期我在阅读您的文献，并尝试复现您的代码过程中受益匪浅。非常感谢您所做的研究工作及分享的代码！在复现过程中，我遇到了一些问题，想向您请教，希望您能给予指点。

首先，在对 TUAB 数据集进行预处理时，我注意到有三个文件总是被归类为错误文件，无法正常处理。具体文件如下：
- aaaaakfo_s004_t000.edf
- aaaaakfo_s005_t000.edf
- aaaaajsl_s010_t000.edf

请问这些问题是否可能是由于数据集本身存在问题，或者有其他可能的原因？

其次，在进行模型微调（finetuning）的过程中，我收到以下警告：
```
Warning: NaN or Inf found in input tensor.
```
我想请教，这个问题是否也与数据集或预处理步骤相关？

最后，在使用您提供的默认参数进行微调时，代码报错了。我尝试安装pretraining中的参数将 `--abs_pos_emb` 参数修改为 `True` ，'--disable_rel_pos_bias' 修改为 ’False‘ ，但不确定是否因为这一改动导致了其他问题。请问该参数是否需要修改？

非常感谢您的时间与帮助，期待您的回复！

**Comment by dyhan316** [comment]
I actually got the same problem (NaN/InF in input tensor). 
How did you manage it? It seems that only a few edf files are like that... should I remove them?

**Comment by ChenHu-ML** [comment]
从我复现的结果，应该是不需要处理这个问题，但我发现，他原文给的fine-tune代码似乎有一些参数不太对。请问一下你是如何处理这些参数的？

---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年11月10日(周日) 上午6:38
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [935963004/LaBraM] 关于数据预处理和finetuning的一些问题 (Issue #45)




 
I actually got the same problem (NaN/InF in input tensor).
 How did you manage it? It seems that only a few edf files are like that... should I remove them?

**Comment by Zhreyu** [comment]
Hi @ChenHu-ML,

I attempted to replicate this experiment as well, but my results on the test set were only at chance level. It seems the model may be overfitting from the start, leading to poor generalization and suboptimal test performance. Could you share the test set scores you achieved in your experiment? Any insights would be greatly appreciated. Thank you!

**Comment by ChenHu-ML** [comment]
@dyhan316 Based on my reproduction results, it shouldn’t be necessary to address this issue. However, I noticed that the fine-tuning code provided in the original paper seems to include some incorrect parameters. How would you handle these adjustments?

**Comment by ChenHu-ML** [comment]
Hi @Zhreyu,
my reproduction results align with the original paper result, but they’re typically achieved within the first epoch.

**Comment by Zhreyu** [comment]
Thanks for the response! I’m seeing nearly perfect accuracy on the training set, but the validation results are all zero, possibly due to class imbalance. Not sure how to resolve this. On the test set, the balanced accuracy remains around 50%. Do you have any suggestions on what might be causing this or areas I should investigate further? Thank you for your help!

**Comment by ChenHu-ML** [comment]
@Zhreyu, You may need to use tensorboard to view the results.

**Comment by Zhreyu** [comment]
Thankyou @ChenHu-ML , Did you use the author's original hyperparameters? if not can you please tell me about the hyper parameters  you used?

**Comment by zeydabadi** [comment]
@ChenHu-ML This is my question too.
Could someone specify the complete hyperparameter configuration required to reproduce the experimental results detailed in the paper?

**Comment by ChenHu-ML** [comment]
@zeydabadi @Zhreyu . I deleted the code about the distributed architecture in the code and modified the following parameters. 

'''
    parser.add_argument('--output_dir', default='',
                        help='path where to save, empty for no saving')

    parser.add_argument('--log_dir', default=None,
                        help='path where to tensorboard log')

    parser.add_argument('--model', default='labram_base_patch200_1600_8k_vocab', type=str, metavar='MODEL',
                        help='Name of model to train')
    parser.add_argument('--rel_pos_bias', action='store_true')
    parser.add_argument('--disable_rel_pos_bias', action='store_true', dest='rel_pos_bias')
    parser.set_defaults(rel_pos_bias=False)
    parser.add_argument('--abs_pos_emb', action='store_true')
    parser.set_defaults(abs_pos_emb=True)
    parser.add_argument('--layer_scale_init_value', default=0.1, type=float, 
                        help="0.1 for base, 1e-5 for large. set 0 to disable layer scale")
'''
However, I only reproduced the experimental results on the TUAB dataset.
![test](https://github.com/user-attachments/assets/502cf87e-dba9-4d74-a875-3f1d27529773)

**Comment by dyhan316** [comment]
Oh i also got similar result on tuab using the given parameters, even despite the nan losses (balan acc of ~0.81)

**Comment by ChenHu-ML** [comment]
@dyhan316. Could you reproduce the results in the TUEV dataset? I seemly cannot get the result. The three metrics in the paper are Non.

**Comment by dyhan316** [comment]
@ChenHu-ML  I think I get a similar result!

Val: Total time: 0:00:02 (0.0525 s / it)
* loss 1.013
Accuracy of the network on the 17179 val EEG: 0.73%
Test:  [ 0/74]  eta: 0:00:43  loss: 0.8640 (0.8640)  accuracy: 0.7500 (0.7500)  balanced_accuracy: 0.5603 (0.5603)  cohen_kappa: 0.5347 (0.5347)  f1_weighted: 0.7615 (0.7615)  time: 0.5847  data: 0.5595  max mem: 1653
Test:  [10/74]  eta: 0:00:05  loss: 0.8503 (0.8146)  accuracy: 0.7708 (0.7746)  balanced_accuracy: 0.5682 (0.5852)  cohen_kappa: 0.5512 (0.5704)  f1_weighted: 0.7912 (0.7889)  time: 0.0809  data: 0.0617  max mem: 1653
Test:  [20/74]  eta: 0:00:03  loss: 0.7468 (0.7925)  accuracy: 0.7812 (0.7867)  balanced_accuracy: 0.5803 (0.5870)  cohen_kappa: 0.5671 (0.5811)  f1_weighted: 0.7968 (0.8004)  time: 0.0358  data: 0.0169  max mem: 1653
Test:  [30/74]  eta: 0:00:02  loss: 0.7428 (0.7887)  accuracy: 0.7917 (0.7866)  balanced_accuracy: 0.5830 (0.5892)  cohen_kappa: 0.5841 (0.5801)  f1_weighted: 0.8160 (0.8026)  time: 0.0351  data: 0.0160  max mem: 1653
Test:  [40/74]  eta: 0:00:01  loss: 0.7643 (0.7804)  accuracy: 0.7917 (0.7879)  balanced_accuracy: 0.5830 (0.5882)  cohen_kappa: 0.5920 (0.5799)  f1_weighted: 0.8090 (0.8052)  time: 0.0335  data: 0.0144  max mem: 1653
Test:  [50/74]  eta: 0:00:01  loss: 0.7690 (0.7902)  accuracy: 0.7917 (0.7853)  balanced_accuracy: 0.6075 (0.5948)  cohen_kappa: 0.5942 (0.5801)  f1_weighted: 0.8090 (0.8025)  time: 0.0401  data: 0.0210  max mem: 1653
Test:  [60/74]  eta: 0:00:00  loss: 0.7690 (0.7868)  accuracy: 0.7812 (0.7864)  balanced_accuracy: 0.5850 (0.5944)  cohen_kappa: 0.5692 (0.5808)  f1_weighted: 0.8021 (0.8036)  time: 0.0440  data: 0.0251  max mem: 1653
Test:  [70/74]  eta: 0:00:00  loss: 0.7675 (0.7868)  accuracy: 0.7812 (0.7859)  balanced_accuracy: 0.5850 (0.5955)  cohen_kappa: 0.5655 (0.5791)  f1_weighted: 0.8021 (0.8033)  time: 0.0458  data: 0.0273  max mem: 1653
Test:  [73/74]  eta: 0:00:00  loss: 0.7137 (0.7791)  accuracy: 0.8021 (0.7882)  balanced_accuracy: 0.5950 (0.5956)  cohen_kappa: 0.5910 (0.5822)  f1_weighted: 0.8165 (0.8050)  time: 0.0358  data: 0.0177  max mem: 1653
Test: Total time: 0:00:03 (0.0472 s / it)
* loss 0.779
Accuracy of the network on the 28305 test EEG: 0.79%
Max accuracy val: 0.75%, max accuracy test: 0.78%
Training time 0:10:55


Do you know how I can view the thing you viewed?

**Comment by zeydabadi** [comment]
@dyhan316 Can you please report your final balanced accuracy on TUAB?
@ChenHu-ML were you able to reproduce the paper results for TUAB?

**Comment by ChenHu-ML** [comment]
@dyhan316 @zeydabadi 

I recently preprocessed the TUAB dataset using the same preprocessing script on two different servers, but the number of processed samples varied by more than 10,000. Even when running the same script twice on the same server, I observed a difference of over 1,000 samples.

Has anyone else encountered this issue? Could this be due to differences in hardware, dependencies, or random shuffling in the code? Any insights would be greatly appreciated!

Thanks in advance for your help! 🙏
